### PR TITLE
add teerex pallet

### PR DESCRIPTION
**Pull request by zTgx**
- [x]  add pallets: sidechain/teeracle/teerex
- [x]  add upstream_commit
- [x]  add script: generate_teerex_upstream_patch

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -8,6 +8,11 @@ members = [
   'pallets/identity-management',
   'pallets/identity-management-mock',
   'pallets/vc-management',
+  'pallets/tee/primitives/teerex',
+  'pallets/tee/sidechain',
+  'pallets/tee/teeracle',
+  'pallets/tee/teerex',
+  'pallets/tee/test-utils',
```
Comment (Kailai-Wang) [inline Cargo.toml]: IMO putting them directly under `pallets/` would be better, reasons:
- to be consistent with other pallets
- these are not tee-specific, some extrinsics should be called from worker, but some are callable by users

The primitives should be put in different pallets, or in the root `primitives` if anything is common enough.

File: .github/workflows/parachain-ci.yml [hunk 2]
```diff
@@ -240,7 +240,7 @@ jobs:
           default: true
 
       - name: Run unittests
-        run: cargo test --locked --release -p pallet-* --lib
+        run: cargo test --features=skip-ias-check --locked --release -p pallet-* --lib
```
Comment (Kailai-Wang) [inline .github/workflows/parachain-ci.yml:243]: this is really strange, running `pallet-sidechain` alone is fine but it fails with wildcard.

Can you try to investigate what happened? Is it related to the file structure (they are in subdirectories of `pallets`/)

**Review by Kailai-Wang — COMMENTED** [review]
Thanks! I've raised two questions

File: .github/workflows/parachain-ci.yml (hunk 2 above)
Comment (Kailai-Wang) [reply .github/workflows/parachain-ci.yml:243]: Do we still need this feature? Or do we actually need two lines, one with `skip-ias-check` feature, one without

**Review by Kailai-Wang — APPROVED** [review]
Looks good, thanks. 

I left a small comment though. Apart from that, the shell script needs adjustment too due to the changed directory structure.

File: .github/workflows/parachain-ci.yml (hunk 2 above)
Comment (zTgx) [reply .github/workflows/parachain-ci.yml:243]: Yes, we need.
> enable `skip-ias-check` to run `identity-management` unit tests and 
> disable to run  `pallet-sidechain` 's unit tests

**Review by Marky-Shi — COMMENTED** [review]
This doesn‘t seem to be the result of executing the benchmark, and needs to be regenerated via benchmark-code.

I ran the benchmark code locally and some errors were encountered, I'll go and try to fix it.

Starting benchmark: pallet_teerex::register_enclave
2022-12-27 10:08:57 panicked at 'Benchmark does not allow the `skip-ias-check` flag.'

**Comment by zTgx** [comment]
> This doesn‘t seem to be the result of executing the benchmark, and needs to be regenerated via benchmark-code.
> 
> I ran the benchmark code locally and some errors were encountered, I'll go and try to fix it.
> 
> Starting benchmark: pallet_teerex::register_enclave 2022-12-27 10:08:57 panicked at 'Benchmark does not allow the `skip-ias-check` flag.'

run benchnark needs to disable `tee-dev` feature

**Comment by Marky-Shi** [comment]
> > This doesn‘t seem to be the result of executing the benchmark, and needs to be regenerated via benchmark-code.
> > I ran the benchmark code locally and some errors were encountered, I'll go and try to fix it.
> > Starting benchmark: pallet_teerex::register_enclave 2022-12-27 10:08:57 panicked at 'Benchmark does not allow the `skip-ias-check` flag.'
> 
> run benchnark needs to disable `tee-dev` feature

By default, tee-dev features are enabled, but it's not certain what the effect of canceling this feature will be. But for now, the existence of tee-dev affects the generation of teerex pallet weight files.

**Comment by Marky-Shi** [comment]
> > > This doesn‘t seem to be the result of executing the benchmark, and needs to be regenerated via benchmark-code.
> > > I ran the benchmark code locally and some errors were encountered, I'll go and try to fix it.
> > > Starting benchmark: pallet_teerex::register_enclave 2022-12-27 10:08:57 panicked at 'Benchmark does not allow the `skip-ias-check` flag.'
> > 
> > 
> > run benchnark needs to disable `tee-dev` feature
> 
> By default, tee-dev features are enabled, but it's not certain what the effect of canceling this feature will be. But for now, the existence of tee-dev affects the generation of teerex pallet weight files.

make build-node-benchmarks 
./scripts/benchmark-weight-local.sh rococo pallet-teerex pallet
./scripts/benchmark-weight-local.sh rococo pallet-teerex runtime

**Pull request merged**